Fix all remaining System.Security.Cryptography.Algorithms test suite crashes.

jkoritzinsky opened the pull request:
With this + PR #48348, the test suite runs to completion and reports the following results:

> Tests run: 2527 Passed: 1285 Inconclusive: 0 Failed: 1227 Ignored: 7

So we've got a ways to go to get things passing, but now we can easily track it at least for this assembly.

jkoritzinsky committed:
Fix all remaining System.Security.Cryptography.Algorithms test suite crashes.

With this + PR #48348, the test suite runs to completion and reports the following results:

> Tests run: 2527 Passed: 1285 Inconclusive: 0 Failed: 1227 Ignored: 7

So we've got a ways to go to get things passing, but now we can easily track it at least for this assembly.

jkoritzinsky commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesGcm.Unix.cs:
```diff
@@ -111,7 +111,7 @@ private void ImportKey(ReadOnlySpan<byte> key)
             if (plaintextBytesWritten != plaintext.Length)
             {
-                Debug.Fail($"GCM decrypt wrote {plaintextBytesWritten} of {plaintext.Length} bytes.");
+                 // Debug.Fail($"GCM decrypt wrote {plaintextBytesWritten} of {plaintext.Length} bytes.");
```
I'll work on trying to isolate and fix this issue, but in the meantime this is the easiest way to keep this `Debug.Fail` method from taking down the test process.

bartonjs replied:
Unlike ECB and CBC, GCM produces one byte of output for one byte of input.  Goodness isn't happening if this assert fails (and tests seemingly shouldn't be passing, unless they got the data from elsewhere having depended on this assert)

bartonjs replied:
If this is needed, please don't merge without creating an issue to put it back.

jkoritzinsky replied:
Oh the tests aren’t passing. But this commenter out assert allows them to run and fail and have XUnit record the issue.

jkoritzinsky replied:
They fail from the exception that immediately follows this assert.

jkoritzinsky replied:
Tracking issue for re-enabling the assert is here: https://github.com/dotnet/runtime/issues/48471

elinor-fung commented:
> we can easily track it

Is this suite in some state where certain classes/groupings of tests pass now (with your other PR)?

I added `System.Security.Cryptography.Algorithms` to #45740. I was thinking we could break down the suites listed there into their test classes/collections (or whatever break-down makes sense for each suite) and mark things off as we get them working. Thoughts?

elinor-fung commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_evp_cipher.c:
```diff
@@ -360,12 +360,17 @@ int32_t CryptoNative_EvpCipherSetGcmTag(CipherCtx* ctx, uint8_t* tag, int32_t ta
+    jbyteArray inDataBytes = (*env)->NewByteArray(env, tagLength);
+    (*env)->SetByteArrayRegion(env, inDataBytes, 0, tagLength, (jbyte*)tag);
+    jbyteArray outDataBytes = (jbyteArray)(*env)->CallObjectMethod(env, ctx->cipher, g_cipherUpdateMethod, inDataBytes);
+    (*env)->DeleteLocalRef(env, outDataBytes);
```
Delete `inDataBytes` too

elinor-fung commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_evp_cipher.c:
```diff
@@ -360,12 +360,17 @@ int32_t CryptoNative_EvpCipherSetGcmTag(CipherCtx* ctx, uint8_t* tag, int32_t ta
+    JNIEnv* env = GetJNIEnv();
+    jbyteArray inDataBytes = (*env)->NewByteArray(env, tagLength);
+    (*env)->SetByteArrayRegion(env, inDataBytes, 0, tagLength, (jbyte*)tag);
+    jbyteArray outDataBytes = (jbyteArray)(*env)->CallObjectMethod(env, ctx->cipher, g_cipherUpdateMethod, inDataBytes);
```
`CheckJNIExceptions`?

jkoritzinsky committed:
PR Feedback.

jkoritzinsky commented:
There's some suites that fully pass and others that are partial. I'll try to pull together a list for the issue based on this PR.

jkoritzinsky commented:
@elinor-fung I've updated the table with the list of test suites and added a link to the program I used to get the list so we can use it for the other assemblies and as we complete the test suites.
